### PR TITLE
Feature: Webapp can have engine connectivity

### LINKED ISSUE #54: Make converter engine-connective
### Is your feature request related to a problem? Please describe.

When converting diagrams, most of them should be automated at this time.

These diagrams will reside in the engine and would have to be downloaded and uploaded to the converter.

### Describe the solution you'd like

The solution would be a configurable connection to the process engine which would allow polling of all process diagrams, conversion and presentation of the results.

### Describe alternatives you've considered

Using local checkouts of a gh repo could also be handy.

### Additional context

This becomes a lot easier when #53 is done.

### PR DESCRIPTION
**Pull request by jonathanlukas**
## Description
Currently, the webapp can only convert diagrams directly provided to it.

Now, it can be used to extract process definitions directly from the process engine REST API, convert it and download it.

## Additional context
Closes #54 

## Testing your changes
There were no tests added.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an existing open issue)
- [x] New feature (non-breaking change which adds functionality to an extension)
- [ ] Breaking change (fix or feature that would cause existing functionality of an extension to change)
- [ ] Documentation update (changes made to an existing piece of documentation)

## Checklist:
- [x] My code adheres to the syntax used by this extension.
- [x] My pull request requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **Camunda Community Hub** documentation.
- [x] I have read the **Pull Request Process** documentation.
- [ ] I have added or suggested tests to cover my changes suggested in this pull request.
- [x] All new and existing CI/CD tests passed.
- [x] I will /assign myself this issue, and add the relevant [issue labels] to it if they are not automatically generated by Probot.
- [x] I will tag @camunda-community-hub/devrel in a new comment on this issue if 30 days have passed since my pull request was opened and I have not received a response from the extension's maintainer.

### CODE REVIEW

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/dto/ProcessDefinitionDiagramDto.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.camunda.community.converter.webapp.dto;
+
+public class ProcessDefinitionDiagramDto {
+  private String bpmn20Xml;
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/dto/ProcessDefinitionDiagramDto.java:4]: Why do you use an Object as a wrapper around a String?

I don't see any value here.

Comment (jonathanlukas) [reply backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/dto/ProcessDefinitionDiagramDto.java:4]: This is the structure the Rest API provides when fetching the XML of a process definition. I cannot change this...

**Comment by ingorichtsmeier** [comment]
I think that this manifest stuff with the engine connection is too complicated.

My assumption is, that the web app is installed as a public service on the internet, where you can check and convert BPMN files stored on your hard drive.

As a power user, you may use the command line client to convert all (or many) files that you find either on your hard drive or in a running Camunda Platform 7 installation.

At the moment, you configure the engine URL in the application.yml and run the web app in a local installation. This is too much effort if you can get the same result from a public service.

It is more reasonable to enter the connection properties on a command line, only available on your local computer, without providing them to a website.

File: backend-diagram-converter/webapp/src/main/resources/static/index.html [hunk 2]
```diff
@@ -123,14 +126,17 @@ <h2 class="accordion-header" id="headingOne">
...
-            return null;
-        }
+    const createFormData = async () => {
+        const manifest = await manifestAwaiter;
         const formData = new FormData();
-        for (let i = 0; i < fileUpload.files.length; i++) {
-            formData.append("files", fileUpload.files[i], fileUpload.files[i].name)
+        if (!manifest.engineEnabled) {
+            if (fileUpload.files.length === 0) {
+                alert("Please select a .bpmn file");
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/resources/static/index.html:134]: Should be "Please select a BPMN file". The file suffix doesn't matter.

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/properties/ProcessEngineClientProperties.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
+package org.camunda.community.converter.webapp.properties;
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/properties/ProcessEngineClientProperties.java:1]: This class should be moved into the cli module.

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/dto/ConverterManifestDto.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package org.camunda.community.converter.webapp.dto;
+
+public class ConverterManifestDto {
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/dto/ConverterManifestDto.java:3]: This is overengineered.

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/ProcessEngineService.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
+package org.camunda.community.converter.webapp;
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/ProcessEngineService.java:1]: Should be moved into the cli module

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/ManifestController.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.camunda.community.converter.webapp.dto.ConverterManifestDto;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class ManifestController {
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/ManifestController.java:11]: This is overengineered

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/ConverterController.java [hunk 7]
```diff
@@ -28,64 +30,97 @@
...
+                      BpmnDiagramCheckResult result = new BpmnDiagramCheckResult();
+                      result.setFilename(bpmnFile.getOriginalFilename());
+                      return result;
+                    }
+                  })
+              .collect(Collectors.toList());
+      return ResponseEntity.ok(results);
+    }
+    if (bpmnFiles == null && engineEnabled) {
+      return ResponseEntity.ok(bpmnConverter.checkFromEngine(false));
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/ConverterController.java:64]: Don't read the file from the engine in the web app, it should be done in the cli module.

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/ConverterController.java [hunk 8]
```diff
@@ -28,64 +30,97 @@
...
       }
-    } catch (IOException e) {
-      return ResponseEntity.internalServerError().body(e.getMessage());
     }
-    if (exceptions.isEmpty()) {
+    if (bpmnFiles == null && engineEnabled) {
+      ByteArrayOutputStream bo = new ByteArrayOutputStream();
+      try (ZipOutputStream out = new ZipOutputStream(bo)) {
+        for (Entry<String, BpmnModelInstance> bpmnFile :
+            bpmnConverter.convertFromEngine(appendDocumentation).entrySet()) {
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/ConverterController.java:110]: Don't read the file from the engine in the web app, it should be done in the cli module.

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/BpmnConverterService.java [hunk 9]
```diff
@@ -27,9 +36,37 @@ public BpmnDiagramCheckResult check(
     return bpmnConverter.check(filename, modelInstance, appendDocumentation);
   }
 
-  public String printXml(DomDocument document, boolean b) {
+  public Map<String, BpmnModelInstance> convertFromEngine(boolean appendDocumentation) {
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/BpmnConverterService.java:39]: This should be done in the cli module.

File: backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/BpmnConverterService.java [hunk 10]
```diff
@@ -27,9 +36,37 @@ public BpmnDiagramCheckResult check(
...
+                e -> {
+                  BpmnModelInstance modelInstance =
+                      Bpmn.readModelFromStream(streamForString(e.getValue()));
+                  convert(modelInstance, appendDocumentation);
+                  return modelInstance;
+                }));
+  }
+
+  public List<BpmnDiagramCheckResult> checkFromEngine(boolean appendDocumentation) {
+    return processEngineService.getAllLatestBpmnXml().entrySet().stream()
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/src/main/java/org/camunda/community/converter/webapp/BpmnConverterService.java:53]: This should be done in the cli module.

File: backend-diagram-converter/webapp/README.md [hunk 11]
```diff
@@ -6,28 +6,46 @@ This webapp uses the diagram converter to check and convert diagrams. It allows
 
 ### Frontend
 
-The frontend is self-explanatory. You upload .bpmn Files and then click either **Check** or **Convert and download**.
+The frontend is self-explanatory. In case no engine connection is configured, you upload .bpmn Files and then click either **Check** or **Convert and download**.
```
Comment (ingorichtsmeier) [inline backend-diagram-converter/webapp/README.md:9]: All changes should be moved into the cli module.

**Review by ingorichtsmeier — CHANGES_REQUESTED** [review]
Most of these features should be moved from the web app into the cli module.

**Comment by jonathanlukas** [comment]
I think these changed are perfect here. They are part of a migration story which will end up in migrating process instances.

We could think about moving the engine connection to core and also implement it with CLI.

**Comment by ingorichtsmeier** [comment]
@jonathanlukas, 

could you please elaborate on this: 

> They are part of a migration story which will end up in migrating process instances.

Imagine there are many installations of the conversion web app around the globe on several servers in different companies. It becomes very tedious to make all admins to upgrade them in case we offer better migration support.

It would be easier with a command line installation and a central web app installation.

**Comment by jonathanlukas** [comment]
@ingorichtsmeier This is exactly what I had in mind:

The webapp can be used to monitor the connected process engine, identifying potential problems when migrating a process.

Then, it will be able to connect it Camunda 8 (Zeebe and Operate).

From Operate it can fetch process definition data and possibly match by bpmn process id. It can then offer a migration of running process instances.

Et voila, automatic migration.

**Pull request closed without merging**